Reduce ckb CPU usage when mute state polling is not necessary

jbendig opened the pull request:
Selecting a mode with no animations and minimizing the GUI to the system tray constantly uses ~4.3%-6.6% CPU according to `top`. This minor change reduces it to ~0.3%. I didn't notice anything broken from the change but I only tested it with a STRAFE keyboard on Linux.

pjcv commented:
In testing this I found I wasn't able to change modes, lighting, etc. on my mouse.

It's my understanding that the Kb::frameUpdate is responsible for updating the device for essentially any purpose, including lighting, DPI settings, and indicator lights, all of which must happen even without animations on. If I follow the logic here correctly, none of that will happen if the animation list is empty.

jbendig commented:
That's definitely not right! 😄 

I think profiling lead me astray when I originally worked on this. After some more poking around, I found that polling for the mute state is the biggest use of CPU time. I added a couple more commits  that help with this.

1. In the Linux `getMuteState` function, don't run `pacmd` again for 30 seconds if it fails to run. On my Fedora 26 system, `pacmd` is provided by the `pulseaudio-utils` package which isn't necessarily installed. This approach means it'll work automatically if installed while ckb is running. I had to insert "set -o pipefail" into the command being run because otherwise the exit code from running `sh` was always 0. I'm not sure if this will cause problems on platforms where `sh` is not `bash`.
2. Explicitly disable the mute indicator on STRAFE keyboards because they don't have a light for it. Previously, the checkbox was hidden but left enabled.

Either commit reduces the CPU usage down to ~1.3% on my system. It's not as drastic of reduction but it's good enough to scratch my itch.

pjcv commented:
I also found that the mute state polling is quite expensive (see #166), but in my case pacmd is actually available. I agree we should probably back off if the command fails, but I think it would be worth considering doing something more drastic here. I wonder if more latency would be acceptable.

Also I think the indicator enabling logic needs a closer look - the mute indicator is running even though I am using a mouse. The indicators are always assumed to have a specific key name, perhaps we can check the device's keys and disable those which don't exist.

jbendig commented:
Do you have a suggested interval in mind? I imagine that the light doesn't have to be as responsive since the audio itself gives the user feedback when the button is pressed.

I'll poke around some more and see about using the key map to determine which indicators should be enabled. It looks like the widgets in kperfwidget should be selectively hidden this way instead of by model too. Does this seem reasonable?

pjcv commented:
Maybe 1/10th of a second? Still seems like a band-aid...

If I understand correctly, right now there are three cases for indicators: k95, strafe, and other. And the strafe case just removes the mute option. I think it would make sense to replace the strafe conditional with one that checks for an actual mute indicator in the keymap.

That being said there's probably no reason why we couldn't extend the keymap checking for all indicators, but checking each one independently might be a bit verbose and unnecessary if they're just specific to the one model anyway.

jbendig commented:
> Maybe 1/10th of a second? Still seems like a band-aid...

Maybe use D-Bus or libpulse to listen for the change asynchronously?

pjcv commented:
That would be cool, hopefully it would be less resource-intensive...

jbendig commented:
I'll do some experimenting and see.

jbendig committed:
Change getMuteState() on Linux to use libpulse.

The mute state is now updated asynchronously so calling this function is much cheaper.

jbendig commented:
I tried using libpulse and it appears to work well. The implementation is a bit more complicated though. There is a background thread running to manage events. It's possible to avoid this thread by polling manually every time getMuteState() is called or by assuming Qt was compiled with glib.

I think all of the edge cases are handled but I could use some help testing. Since my keyboard does not have a mute button, my testing was limited.

pjcv commented:
Awesome, this works for me, and it seems the CPU usage is much less than before.

Note I wasn't able to compile without adding c++11 due to the use of the nullptr keyword.

jbendig commented:
I can replace nullptr with NULL easily enough.

For whatever it's worth, Qt5 >= 5.7 requires C++11. Is there a minimum version of Qt5 supported by ckb? The README currently recommends 5.9.

ghost commented:
@jbendig

technically, the min version is 5.2.

> I can replace nullptr with NULL easily enough.

I'd say don't bother and instead add the needed flag. C++11 has been supported by compilers for half a decade now, so it seems to me an easy requirement.

ghost commented:
Oh, actually, `-gnu++11` is already passed for the GUI, it requires 24e8f01.

Also, as a side note, IIRC from the cfe-dev mailing list, clang++ 6.0 will ship with the default of c++11 instead of c++98. Don't know about g++ though.

jbendig commented:
> Oh, actually, -gnu++11 is already passed for the GUI, it requires 24e8f01.

I won't worry about it then. I'll just make sure to rebase once I'm ready to clean-up the PR.

> Also, as a side note, IIRC from the cfe-dev mailing list, clang++ 6.0 will ship with the default of c++11 instead of c++98. Don't know about g++ though.

g++ skipped c++11 and made c++14 default as of GCC 6. https://gcc.gnu.org/gcc-6/changes.html

jbendig commented:
I'm going to remove 51f3054 from this PR. There's no reason for a STRAFE specific exception now that getMuteState() is so much cheaper to call. I still might be willing to make the GUI dynamically hide indicators based on the keymap.

The Pulse Audio change reduces the CPU usage when no animations are playing to ~1.3%. I have a patch that reduces this down to 0.0-0.3% but I'll submit it as a separate PR so it's easier to test this one.

jbendig commented:
PR should be ready for review now.

I thought 24e8f01 was a newer commit in newdev but it's only in master so I didn't need to rebase.

tatokis commented on src/ckb/media_linux.cpp:
```diff
@@ -1,34 +1,142 @@
+    //Connect to the local Pulse Audio server. It's usually running but a
+    //reconnect is attempted periodically whenever the connection fails or is
+    //terminated.
+    if(paContext == nullptr && QDateTime::currentMSecsSinceEpoch() >= reconnectTime){
```
Minor issue, but it would be nice to get rid of the warning.
```
media_linux.cpp: In function ‘muteState getMuteState()’:
media_linux.cpp:125:68: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     if(paContext == nullptr && QDateTime::currentMSecsSinceEpoch() >= reconnectTime){
                                                                    ^
```

jbendig replied:
It should be fixed now. I tested with both g++  and clang.

tatokis replied:
Excellent, thank you!

jbendig committed:
Fix compile warning on Linux when building with Pulse Audio

ghost commented:
The CPU usage (according to `top`) [for my i3] went down from 7.3% to 1.3% when idle and without animations. Thanks a lot! Merging.

ghost commented:
Also, just not to misinform you: I found the [cfe-dev thread](http://lists.llvm.org/pipermail/cfe-dev/2017-February/052820.html) I was talking about and it is not yet decided on whether clang 6 will ship with 11'th dialect, 14'th dialect or won't change anything.
